### PR TITLE
[CPU] Unique operation implementation.

### PR DESCRIPTION
**Pull request by nshchego**
### Details:
 - *Add new operation "Unique-10" to IE CPU Plugin*

### Tickets:
 - *93675*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@chenhu-wang could you please review the PR? It is targeting 2022, so the review is requires asap.
Please skip ngraph part this is a copy of related changes from another PR.

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        } else if (dataPrecision == Precision::I32) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::I32>::value_type>();
+        } else if (dataPrecision == Precision::I8) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::I8>::value_type>();
+        } else if (dataPrecision == Precision::U8) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::U8>::value_type>();
+        }
+
+        const size_t srcLen = getParentEdgeAt(IN_DATA)->getMemoryPtr()->GetSize() / dataPrecision.size();
+        redefineOutputMemory({ {uniqueLen}, {uniqueLen}, {srcLen}, {uniqueLen}});
```
Comment (chenhu-wang) [inline src/plugins/intel_cpu/src/nodes/unique.cpp:121]: seems like redefineOutputMemory() happens after all computation finished, which means the computation result stored in orignal child edge is lost. Is there any copy for the computation result to this new allocated memory?

**Review by chenhu-wang — CHANGES_REQUESTED** [review]

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        } else if (dataPrecision == Precision::I8) {
+            uniqueLen = slicedTensorExec<PrecisionTrait<Precision::I8>::value_type>();
+        } else if (dataPrecision == Precision::U8) {
+            uniqueLen = slicedTensorExec<PrecisionTrait<Precision::U8>::value_type>();
+        }
+
+        auto dstDataShape = getParentEdgeAt(IN_DATA)->getMemoryPtr()->getStaticDims();
+        const size_t srcLen = dstDataShape[axis];
+        dstDataShape[axis] = uniqueLen;
+        redefineOutputMemory({ dstDataShape, {uniqueLen}, {srcLen}, {uniqueLen}});
```
Comment (chenhu-wang) [inline src/plugins/intel_cpu/src/nodes/unique.cpp:137]: the same, need copy the result to new memory, or redefineOutputMemory should be before execution of store result.

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+void Unique::executeDynamicImpl(dnnl::stream strm) {
+    execute(strm);
+}
+
+template <typename T>
+size_t Unique::flattenTensorExec() {
+    const T* srcDataPtr = reinterpret_cast<const T*>(getParentEdgeAt(IN_DATA)->getMemoryPtr()->GetPtr());
+    const int64_t inputLen = getParentEdgeAt(IN_DATA)->getMemoryPtr()->GetSize() / sizeof(T);
+    T* uniqueData = reinterpret_cast<T*>(getChildEdgesAtPort(UNIQUE_DATA)[0]->getMemoryPtr()->GetPtr());
+    int *firstPtr, *inToOutPtr, *occurPtr;
```
Comment (chenhu-wang) [inline src/plugins/intel_cpu/src/nodes/unique.cpp:150]: uninitialized variance is used, which result "warning as error" when building.

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+                 if (definedOutputs[FIRST_UNIQUE_IDX]) {
+                     firstPtr[uniqLen] = i;
+                 }
+                 if (definedOutputs[INPUT_TO_UNIQ_IDX]) {
+                     inToOutPtr[i] = j;
+                 }
+
+                 uniqLen++;
+             } else {
+                 if (definedOutputs[INPUT_TO_UNIQ_IDX]) {
```
Comment (chenhu-wang) [inline src/plugins/intel_cpu/src/nodes/unique.cpp:238]: minor: the same for both condition, maybe could be done out of this condition.

File: src/core/reference/include/ngraph/runtime/reference/unique.hpp [hunk 5]
```diff
@@ -113,16 +113,24 @@ UniqueElements<Index_t, Count_t> find_unique_elements(const Data_t* data,
...
     };
 
+    int64_t axisVal = 0;
+    if (axis) {
+        axisVal = *axis;
+        if (axisVal < 0) {
+            axisVal += data_shape.size();
+        }
+    }
+
```
Comment (mitruska) [inline src/core/reference/include/ngraph/runtime/reference/unique.hpp:123]: @tomdol Looks like negative axis should be covered by Unique reference implementation and the reference tests as well.

**Comment by mitruska** [comment]
@nshchego Related PR https://github.com/openvinotoolkit/openvino/pull/14229 has been merged into the master, so please update this branch to avoid further conflicts.

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::FP32>::value_type>();
+        } else if (dataPrecision == Precision::I32) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::I32>::value_type>();
+        } else if (dataPrecision == Precision::I8) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::I8>::value_type>();
+        } else if (dataPrecision == Precision::U8) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::U8>::value_type>();
+        }
+
+        const size_t srcLen = getParentEdgeAt(IN_DATA)->getMemoryPtr()->GetSize() / dataPrecision.size();
```
[the remark this answers is not in the document]
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/unique.cpp]: Done

File: src/plugins/intel_cpu/src/nodes/unique.cpp (hunk 3 above)
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/unique.cpp]: Done

File: src/plugins/intel_cpu/src/nodes/unique.cpp (hunk 4 above)
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/unique.cpp]: Fixed

File: src/plugins/intel_cpu/src/nodes/unique.cpp (hunk 1 above)
Comment (chenhu-wang) [reply src/plugins/intel_cpu/src/nodes/unique.cpp:121]: 1. ok, maybe a special case need double check. Do we support dynamic input shape for this op? If yes and initial shape is undefined(low/upper bound is not provide), then there is possibility that real reallocation happens. for example, first infer input value is [1, 1, 1], second infer input value is [1,2,3,4,5]. second uniqueLen is even larger than first srcLen(preallocate with this size), which need reallocation. Am I right?
2. I see when shape is the same, there will be no reallocation in https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/node.cpp#L560-L561. could you please help me understand where the reallocation will be skipped when new shape is smaller than before.

Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/unique.cpp:121]: 1. That is right point for dynamic shapes. I lost this initialization.
2. https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/cpu_memory.cpp#L185

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+    return uniqLen;
+}
+
+template <typename T>
+size_t Unique::slicedTensorExec() {
+    const T* srcDataPtr = reinterpret_cast<const T*>(getParentEdgeAt(IN_DATA)->getMemoryPtr()->GetPtr());
+    const auto inputLen = getParentEdgeAt(IN_DATA)->getMemoryPtr()->GetSize() / sizeof(T);
+    T* uniqueData = reinterpret_cast<T*>(getChildEdgesAtPort(UNIQUE_DATA)[0]->getMemoryPtr()->GetPtr());
+    int *firstPtr, *inToOutPtr, *occurPtr;
```
[the remark this answers is not in the document]
Comment (chenhu-wang) [reply src/plugins/intel_cpu/src/nodes/unique.cpp:263]: this is a reason of failed CI

**Comment by chenhu-wang** [comment]
In general LGTM, need fix CI issues.

**Review by chenhu-wang — APPROVED** [review]

File: src/plugins/intel_cpu/src/nodes/unique.cpp (hunk 1 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/src/nodes/unique.cpp:122]: That's completely unsafe way of memory redefinition. You are exploiting knowledge of current memory reallocation behavior, which in fact depends on underlaying memory manager. Actual memory manager - not under Node control, so it's change can make this logic invalid. 
So @chenhu-wang  is right: you need to use temporary memory object and then copy data to dst.

File: src/plugins/intel_cpu/src/nodes/unique.cpp [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+    if (getSelectedPrimitiveDescriptor() == nullptr) {
+        THROW_ERROR << " has unidentified preferable primitive descriptor.";
+    }
+}
+
+void Unique::execute(dnnl::stream strm) {
+    if (flattened) {
+        size_t uniqueLen = 1lu;
+        if (dataPrecision == Precision::FP32) {
+            uniqueLen = flattenTensorExec<PrecisionTrait<Precision::FP32>::value_type>();
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/unique.cpp:112]: Please support conditional compilation feature both for `flattenTensorExec` and `slicedTensorExec` via OV_SWITCH

File: src/plugins/intel_cpu/src/nodes/unique.cpp (hunk 1 above)
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/unique.cpp]: Fixed

File: src/plugins/intel_cpu/src/nodes/unique.cpp (hunk 8 above)
Comment (nshchego) [reply src/plugins/intel_cpu/src/nodes/unique.cpp]: Done

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**